### PR TITLE
26 use runtime managed erc20 tokens in tests

### LINKED ISSUE #26: Use runtime managed ERC20 tokens in tests
The Nabla test use a `MockERC20` contract. These contracts are an extension of the default open zeppelin ERC20 contracts with additional mint and burn functions.

These contracts should be replaced with the `ERC20Wrapper` contracts defined in the `pendulum-ink-wrapper`. However, this contract does not have mint and burn functions and the chain extensions in the foucoco-standalone chain doesn't provide this functionality either.

## TODO
* add mint and burn functions to the chain extension of the [foucoco-standalone](https://github.com/pendulum-chain/foucoco-standalone) chain
* add a new `TestableERC20Wrapper` contract to the [pendulum-ink-wrapper](https://github.com/pendulum-chain/pendulum-ink-wrapper) repository that has burn and mint functions and that uses the chain extensions
* replace the `MockERC20` definition in `nabla/config.json` by referring to the `TestableERC20Wrapper` contract

### PR DESCRIPTION
**Pull request by gianfra-t**
Issue: #26.

Replaced the dependency on MockERC20 for TestableERC20 contract that interacts with the standalone chain during tests.

The tests were modified accordingly to work in this manner. Specifically cleaning token balances before execution of the tests and modifying time delay for swapExecution. Also added a test specific to the TestableERC20Wrapper.

The corresponding modifications for foucococo-standalone and pendulum-wrappers for this to work were made [here](https://github.com/pendulum-chain/foucoco-standalone/pull/5) and [here](https://github.com/pendulum-chain/pendulum-ink-wrapper/pull/34) respectively.

### CODE REVIEW

File: nabla/test/TestableERC20Tests.ts
Comment (ebma) [inline nabla/test/TestableERC20Tests.ts]: We could rename these assets to `assetNative` and `assetToken` or something. Would make the tests below a little easier to read.

File: nabla/test/TestableERC20Tests.ts
Comment (ebma) [inline nabla/test/TestableERC20Tests.ts]: Should we remove this function or do we expect to use it in the future?

File: nabla/test/backstop.ts
Comment (ebma) [inline nabla/test/backstop.ts]: With the constructor arguments, this does not actually represent `native`, does it? Let's either rename the token or set it up to really 'be' `native`

File: nabla/test/swaps.ts
Comment (ebma) [inline nabla/test/swaps.ts:30]: Is there a reason why the `deadlineTime` defined here is different to the places you replace it with? Why is the `+10` needed?

File: nabla/test/backstop.ts
Comment (ebma) [inline nabla/test/backstop.ts:129]: Why do we have to `burn` before we `mint`?

**Review by ebma — COMMENTED** [review]
Looks good to me overall but left some comments.

File: nabla/test/backstop.ts
Comment (ebma) [reply nabla/test/backstop.ts:129]: If this is related to 

> Specifically cleaning token balances before execution of the tests and modifying time delay for swapExecution.

Maybe we should find a clearer way to do this? How about a teardown function, or we add cleanup-related code to the end of each test? Otherwise we should maybe at least add a comment that explains why we burn before we mint.

File: nabla/test/TestableERC20Tests.ts
Comment (gianfra-t) [reply nabla/test/TestableERC20Tests.ts]: Agreed!

File: nabla/test/TestableERC20Tests.ts
Comment (gianfra-t) [reply nabla/test/TestableERC20Tests.ts]: I will remove it. It is probably not going to be used in this file anytime soon.

File: nabla/test/backstop.ts
Comment (gianfra-t) [reply nabla/test/backstop.ts]: Before, the tests were using at all instance the MockERC20 which is also non-native. 
I decided to use XCM(1) given that in `foucoco-standalone` it does not start with a balance in genesis.

I will change the name back to Test token x to avoid this confusion.

File: nabla/test/swaps.ts
Comment (gianfra-t) [reply nabla/test/swaps.ts:30]: This is perhaps the only logic change in the test itself. I do not know the particular reason to set it to one value or the other, I assumed that the logic of the test assumes that deadline must not pass.

Before the deadline was set to `now()`, but given that we are now interacting with the chain differently, sometimes this test used to fail, while other times it did not.
 
Even though 10 seconds is totally arbitrary, I don't know a reason to set it to less (or more) since we are just trying to test the case when we are inside the deadline.

File: nabla/test/backstop.ts
Comment (gianfra-t) [reply nabla/test/backstop.ts:129]: Indeed this is just to ensure that the balance of the tokens for the relevant accounts is 0 before the mint and at the end of this `setUp()` function we have only the expected minted amount.

I will add some comments to clarify this, but I don't think we should add another function since we are already in the `setUp()` function and this is the cleanup for a test that came before (if any), let me know it makes sense @ebma.  

BTW, it is true that perhaps a teardown at the end also makes sense, but given that the tests stops if there is an error, as of now I don't think there is a way to actually make sure that the teardown actually runs. That's why I think for now cleaning just in case in the setUp makes more sense.

File: src/actions/compileContract.ts
Comment (ebma) [inline src/actions/compileContract.ts]: Was this change intended?

File: nabla/test/TestableERC20Tests.ts
Comment (ebma) [inline nabla/test/TestableERC20Tests.ts]: Nit: can we remove these?

File: nabla/test/backstop.ts
Comment (ebma) [reply nabla/test/backstop.ts:129]: Right, I see your point. Although we don't support 'teardown' functions yet, we could just try to execute them (if any) in a new `finally {}` block at the end of the try-catch [here](https://github.com/pendulum-chain/wasm-deploy/blob/ec682a08f06673e1db5b2f637e13c87d7447d482/src/commands/test.ts#L411). This way it will be executed in any case. 

I'm also fine with your existing implementation actually, the only thing that bothers me a bit is that the comment reasoning why we need to burn before minting is only included in one file. And I see why since copy-pasting it into any file is annoying. But with a teardown function for each test, it might be self-explanatory enough that we don't even need a comment in the first place. WDYT?

File: nabla/test/swaps.ts
Comment (ebma) [reply nabla/test/swaps.ts:30]: Sure, makes sense. I'm fine with keeping it like this, was just curious if there is any specific reason for this 👍

File: src/actions/compileContract.ts
Comment (gianfra-t) [reply src/actions/compileContract.ts]: Sorry that is a mistake. Every time it runs in local I need to point to my local solang. Will delete now.

File: nabla/test/backstop.ts
Comment (gianfra-t) [reply nabla/test/backstop.ts:129]: Yes I agree that would be the better approach. Could we perhaps open a separate issue and wait for this one? Since it would be touching a separate part of the code logic not just the tests but also how they run.
In the meantime if you are okay with it I can just copy the comment to all the setups so we don't forget what is going on.

File: nabla/test/swaps.ts
Comment (gianfra-t) [reply nabla/test/swaps.ts:30]: I agree that if the intention of the test is to not test the deadline, it should be more obvious and have even more time ("tend to infinity"), but I am not super sure about this one!

File: nabla/test/backstop.ts
Comment (ebma) [reply nabla/test/backstop.ts:129]: Sounds good 👍 could you create that follow-up ticket?

Comment (gianfra-t) [reply nabla/test/backstop.ts:129]: Sure!

**Review by ebma — APPROVED** [review]
LGTM

**Comment by ebma** [comment]
@gianfra-t you can merge this now if you don't want to add anything. Usually, the author of the PR merges it once approved, because sometimes it might be that the author still wants to add something after it got approved by someone else which would be missed if the reviewer merges it.

File: nabla/test/swaps.ts
Comment (TorstenStueber) [reply nabla/test/swaps.ts:30]: I follow up much later here: in the forge test runner for Ethereum you have total control over time, however, in Foucoco standalone this is not the case (yet). So there will always be some time passing between this line in the test and the actual smart contract function running. Therefore it is reasonable to add a little buffer here.

**Pull request merged**
